File: resolve_silence_cut/audio_extract.py
from __future__ import annotations

import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class ExtractedAudio:
    wav_path: str
    duration_seconds: float


def extract_clip_slice_to_wav(
    file_path: str,
    source_start_frame: int,
    source_end_frame: int,
    fps: float,
) -> ExtractedAudio:
    if source_end_frame <= source_start_frame:
        raise ValueError("Invalid source frame bounds for extraction.")

    start_sec = source_start_frame / fps
    duration_sec = (source_end_frame - source_start_frame) / fps
    if duration_sec <= 0:
        raise ValueError("Clip duration is zero during extraction.")

    temp_dir = tempfile.mkdtemp(prefix="resolve_silence_cut_")
    wav_path = str(Path(temp_dir) / "clip.wav")

    cmd = [
        "ffmpeg",
        "-v",
        "error",
        "-ss",
        f"{start_sec:.6f}",
        "-i",
        file_path,
        "-t",
        f"{duration_sec:.6f}",
        "-ac",
        "1",
        "-ar",
        "48000",
        "-vn",
        "-y",
        wav_path,
    ]

    proc = subprocess.run(cmd, capture_output=True, text=True)
    if proc.returncode != 0:
        raise RuntimeError(
            f"ffmpeg extraction failed for '{file_path}': {proc.stderr.strip()}"
        )

    return ExtractedAudio(wav_path=wav_path, duration_seconds=duration_sec)

File: src/core/audioExtract.ts
import { mkdtemp } from "node:fs/promises";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { execa } from "execa";

export interface ExtractedAudio {
  wavPath: string;
  durationSeconds: number;
}

export async function extractClipSliceToWav(
  filePath: string,
  sourceStartFrame: number,
  sourceEndFrame: number,
  fps: number,
): Promise<ExtractedAudio> {
  if (sourceEndFrame <= sourceStartFrame) {
    throw new Error("Invalid source frame bounds for extraction.");
  }

  const startSec = sourceStartFrame / fps;
  const durationSec = (sourceEndFrame - sourceStartFrame) / fps;
  if (durationSec <= 0) {
    throw new Error("Clip duration is zero during extraction.");
  }

  const tempDir = await mkdtemp(join(tmpdir(), "resolve_silence_cut_"));
  const wavPath = join(tempDir, "clip.wav");

  const args = [
    "-v",
    "error",
    "-ss",
    startSec.toFixed(6),
    "-i",
    filePath,
    "-t",
    durationSec.toFixed(6),
    "-ac",
    "1",
    "-ar",
    "48000",
    "-vn",
    "-y",
    wavPath,
  ];

  const result = await execa("ffmpeg", args, { reject: false });
  if (result.exitCode !== 0) {
    const err = (result.stderr || result.stdout || "").toString().trim();
    throw new Error(`ffmpeg extraction failed for '${filePath}': ${err}`);
  }

  return { wavPath, durationSeconds: durationSec };
}
